BUG: scipy v1.15 breaking for pytest when assert-rewrite is on

### Describe your issue.

I don't know if this is a scipy issue or a pytest one, but the most recent version of scipy (1.15) is causing errors with pytest, specifically if assertion rewriting is enabled.  If the `--assert=plain` flag is passed to pytest, no error occurs.

To reproduce: put this in pytest.ini
```ini
[pytest]
python_files = *.py
```

and run pytest on the following python script:

### Reproducing Code Example

```python
from scipy import *
```


### Error message

```shell
========================================================== test session starts ===========================================================
platform darwin -- Python 3.12.4, pytest-8.3.4, pluggy-1.5.0 -- /Users/bbm/mambaforge/envs/refl1d/bin/python3.12
cachedir: .pytest_cache
rootdir: /private/tmp/test
configfile: pytest.ini
plugins: cov-6.0.0, anyio-4.0.0
collected 0 items / 1 error                                                                                                              

================================================================= ERRORS =================================================================
___________________________________________________ ERROR collecting broken_pytest.py ____________________________________________________
broken_pytest.py:3: in <module>
    from scipy import *
<frozen importlib._bootstrap>:1360: in _find_and_load
    ???
<frozen importlib._bootstrap>:1331: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:935: in _load_unlocked
    ???
/Users/bbm/mambaforge/envs/refl1d/lib/python3.12/site-packages/_pytest/assertion/rewrite.py:184: in exec_module
    exec(co, module.__dict__)
/Users/bbm/mambaforge/envs/refl1d/lib/python3.12/site-packages/scipy/ndimage/__init__.py:156: in <module>
    from ._support_alternative_backends import *
<frozen importlib._bootstrap>:1360: in _find_and_load
    ???
<frozen importlib._bootstrap>:1331: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:935: in _load_unlocked
    ???
/Users/bbm/mambaforge/envs/refl1d/lib/python3.12/site-packages/_pytest/assertion/rewrite.py:184: in exec_module
    exec(co, module.__dict__)
/Users/bbm/mambaforge/envs/refl1d/lib/python3.12/site-packages/scipy/ndimage/_support_alternative_backends.py:65: in <module>
    delegator = getattr(_delegators, func_name + "_signature")
E   AttributeError: module 'scipy.ndimage._delegators' has no attribute '@py_builtins_signature'. Did you mean: 'grey_dilation_signature'?
======================================================== short test summary info =========================================================
ERROR broken_pytest.py - AttributeError: module 'scipy.ndimage._delegators' has no attribute '@py_builtins_signature'. Did you mean: 'grey_dilation_signature'?
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================ 1 error in 1.82s ============================================================
```


### SciPy/NumPy/Python version and system information

```shell
1.15.0 2.0.2 sys.version_info(major=3, minor=12, micro=4, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: extraframeworks
    found: true
    include directory: unknown
    lib directory: unknown
    name: Accelerate
    openblas configuration: unknown
    pc file directory: unknown
    version: unknown
  lapack:
    detection method: extraframeworks
    found: true
    include directory: unknown
    lib directory: unknown
    name: Accelerate
    openblas configuration: unknown
    pc file directory: unknown
    version: unknown
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 2.13.6
Compilers:
  c:
    commands: cc
    linker: ld64
    name: clang
    version: 15.0.0
  c++:
    commands: c++
    linker: ld64
    name: clang
    version: 15.0.0
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.0.11
  fortran:
    commands: gfortran
    linker: ld64
    name: gcc
    version: 13.3.0
  pythran:
    include directory: ../../../../../../private/var/folders/95/0ydz4d79163427j3k5crp3fh0000gn/T/pip-build-env-8oen8nl_/overlay/lib/python3.12/site-packages/pythran
    version: 0.17.0
Machine Information:
  build:
    cpu: aarch64
    endian: little
    family: aarch64
    system: darwin
  cross-compiled: false
  host:
    cpu: aarch64
    endian: little
    family: aarch64
    system: darwin
Python Information:
  path: /private/var/folders/95/0ydz4d79163427j3k5crp3fh0000gn/T/cibw-run-ni2eppd2/cp312-macosx_arm64/build/venv/bin/python
  version: '3.12'
```

I can easily reproduce this error via bumps tests, https://github.com/bumps/bumps.
The problem was caught by debian CI tests, e.g. https://ci.debian.net/data/autopkgtest/unstable/amd64/p/python-bumps/58064845/log.gz

The most recent bumps release v0.9.3 triggers the issue, but they seem to have fixed it in HEAD.  v1.0.0a11 triggers the error, while v1.0.0a12 already passes.  Perhaps that gives more clues to what the problem is.

To reproduce:
```
git clone https://github.com/bumps/bumps
cd bumps
git checkout v1.0.0a11
pytest -v bumps/wsolve.py
```

bumps' tests are set up a little peculiarly.  pytest can find them, but they're not specifically designed for use with pytest. 
That said, the test log (for wsolve.py from v1.0.0a11) is
```
$ pytest -v bumps/wsolve.py 
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.13.2, pytest-8.3.4, pluggy-1.5.0 -- /usr/bin/python3
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase(PosixPath('/projects/python/build/test/bumps/.hypothesis/examples'))
Test order randomisation NOT enabled. Enable with --random-order or --random-order-bucket=<bucket_type>
benchmark: 5.1.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /projects/python/build/test/bumps
configfile: pyproject.toml
plugins: remotedata-0.4.1, cov-5.0.0, hypothesis-6.125.2, xdist-3.6.1, arraydiff-0.6.1, filter-subpackage-0.2.0, random-order-1.1.1, xvfb-3.0.0, typeguard-4.4.1, astropy-0.11.0, doctestplus-1.4.0, openfiles-0.6.0, flaky-3.8.1, anyio-4.8.0, mpi-0.6, subtests-0.13.1, astropy-header-0.2.2, timeout-2.3.1, mock-3.14.0, benchmark-5.1.0
collected 2 items                                                                                                                                                                                                 

bumps/wsolve.py::bumps.wsolve PASSED                                                                                                                                                                        [ 50%]
bumps/wsolve.py::test FAILED                                                                                                                                                                                [100%]

==================================================================================================== FAILURES =====================================================================================================
______________________________________________________________________________________________________ test _______________________________________________________________________________________________________

    def test():
        """
        Check that results are correct for a known problem.
        """
        from numpy.testing import assert_array_almost_equal_nulp
        x = np.array([0, 1, 2, 3, 4], 'd')
        y = np.array([2.5, 7.9, 13.9, 21.1, 44.4], 'd')
        dy = np.array([1.7, 2.4, 3.6, 4.8, 6.2], 'd')
        poly = wpolyfit(x, y, dy, 1)
        px = np.array([1.5], 'd')
>       _, pi = poly.pi(px)  # Same y is returend from pi and ci

bumps/wsolve.py:432: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
bumps/wsolve.py:358: in pi
    return self._conf.pi(A, p)
bumps/wsolve.py:211: in pi
    return self._interval(np.asarray(A), p, 1)
bumps/wsolve.py:183: in _interval
    from scipy.stats import t  # lazy import in case scipy not present
<frozen importlib._bootstrap>:1360: in _find_and_load
    ???
<frozen importlib._bootstrap>:1331: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:935: in _load_unlocked
    ???
/usr/lib/python3/dist-packages/_pytest/assertion/rewrite.py:184: in exec_module
    exec(co, module.__dict__)
/usr/lib/python3/dist-packages/scipy/stats/__init__.py:653: in <module>
    from ._mgc import multiscale_graphcorr
<frozen importlib._bootstrap>:1360: in _find_and_load
    ???
<frozen importlib._bootstrap>:1331: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:935: in _load_unlocked
    ???
/usr/lib/python3/dist-packages/_pytest/assertion/rewrite.py:184: in exec_module
    exec(co, module.__dict__)
/usr/lib/python3/dist-packages/scipy/stats/_mgc.py:7: in <module>
    from scipy.ndimage import _measurements
<frozen importlib._bootstrap>:1360: in _find_and_load
    ???
<frozen importlib._bootstrap>:1331: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:935: in _load_unlocked
    ???
/usr/lib/python3/dist-packages/_pytest/assertion/rewrite.py:184: in exec_module
    exec(co, module.__dict__)
/usr/lib/python3/dist-packages/scipy/ndimage/__init__.py:156: in <module>
    from ._support_alternative_backends import *
<frozen importlib._bootstrap>:1360: in _find_and_load
    ???
<frozen importlib._bootstrap>:1331: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:935: in _load_unlocked
    ???
/usr/lib/python3/dist-packages/_pytest/assertion/rewrite.py:184: in exec_module
    exec(co, module.__dict__)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    import functools
    from scipy._lib._array_api import (
        is_cupy, is_jax, scipy_namespace_for, SCIPY_ARRAY_API
    )
    
    import numpy as np
    from ._ndimage_api import *   # noqa: F403
    from . import _ndimage_api
    from . import _delegators
    __all__ = _ndimage_api.__all__
    
    
    MODULE_NAME = 'ndimage'
    
    
    def delegate_xp(delegator, module_name):
        def inner(func):
            @functools.wraps(func)
            def wrapper(*args, **kwds):
                xp = delegator(*args, **kwds)
    
                # try delegating to a cupyx/jax namesake
                if is_cupy(xp):
                    # https://github.com/cupy/cupy/issues/8336
                    import importlib
                    cupyx_module = importlib.import_module(f"cupyx.scipy.{module_name}")
                    cupyx_func = getattr(cupyx_module, func.__name__)
                    return cupyx_func(*args, **kwds)
                elif is_jax(xp) and func.__name__ == "map_coordinates":
                    spx = scipy_namespace_for(xp)
                    jax_module = getattr(spx, module_name)
                    jax_func = getattr(jax_module, func.__name__)
                    return jax_func(*args, **kwds)
                else:
                    # the original function (does all np.asarray internally)
                    # XXX: output arrays
                    result = func(*args, **kwds)
    
                    if isinstance(result, (np.ndarray, np.generic)):
                        # XXX: np.int32->np.array_0D
                        return xp.asarray(result)
                    elif isinstance(result, int):
                        return result
                    elif isinstance(result, dict):
                        # value_indices: result is {np.int64(1): (array(0), array(1))} etc
                        return {
                            k.item(): tuple(xp.asarray(vv) for vv in v)
                            for k,v in result.items()
                        }
                    elif result is None:
                        # inplace operations
                        return result
                    else:
                        # lists/tuples
                        return type(result)(
                            xp.asarray(x) if isinstance(x, np.ndarray) else x
                            for x in result
                        )
            return wrapper
        return inner
    
    # ### decorate ###
    for func_name in _ndimage_api.__all__:
        bare_func = getattr(_ndimage_api, func_name)
>       delegator = getattr(_delegators, func_name + "_signature")
E       AttributeError: module 'scipy.ndimage._delegators' has no attribute '@py_builtins_signature'. Did you mean: 'grey_dilation_signature'?

/usr/lib/python3/dist-packages/scipy/ndimage/_support_alternative_backends.py:65: AttributeError
================================================================================================ warnings summary =================================================================================================
bumps/wsolve.py::bumps.wsolve
bumps/wsolve.py::bumps.wsolve
bumps/wsolve.py::bumps.wsolve
  <doctest bumps.wsolve[6]>:1: DeprecationWarning: Conversion of an array with ndim > 0 to a scalar is deprecated, and will error in future. Ensure you extract a single element from your array before performing this operation. (Deprecated NumPy 1.25.)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html

---------- coverage: platform linux, python 3.13.2-final-0 -----------
Name                                          Stmts   Miss  Cover
-----------------------------------------------------------------
bumps/__init__.py                                11      4    64%
bumps/bounds.py                                 373    373     0%
bumps/bspline.py                                201    201     0%
bumps/cheby.py                                   29     29     0%
bumps/cli.py                                    334    334     0%
bumps/curve.py                                  237    237     0%
bumps/data.py                                   103    103     0%
bumps/dream/__init__.py                           2      2     0%
bumps/dream/acr.py                               15     15     0%
bumps/dream/bounds.py                           130    130     0%
bumps/dream/build_compiled.py                    40     40     0%
bumps/dream/compiled.py                           4      4     0%
bumps/dream/convergence.py                       86     86     0%
bumps/dream/core.py                             171    171     0%
bumps/dream/corrplot.py                         111    111     0%
bumps/dream/crossover.py                         58     58     0%
bumps/dream/diffev.py                           112    112     0%
bumps/dream/digits.py                           103    103     0%
bumps/dream/entropy.py                          471    471     0%
bumps/dream/exppow.py                            20     20     0%
bumps/dream/formatnum.py                        281    281     0%
bumps/dream/gelman.py                            28     28     0%
bumps/dream/geweke.py                            26     26     0%
bumps/dream/initpop.py                           34     34     0%
bumps/dream/ksmirnov.py                          12     12     0%
bumps/dream/mahal.py                             21     21     0%
bumps/dream/matlab.py                           106    106     0%
bumps/dream/metropolis.py                        39     39     0%
bumps/dream/model.py                             65     65     0%
bumps/dream/outliers.py                          71     71     0%
bumps/dream/parcoord.py                          56     56     0%
bumps/dream/state.py                            565    565     0%
bumps/dream/stats.py                            112    112     0%
bumps/dream/tile.py                              18     18     0%
bumps/dream/util.py                              60     60     0%
bumps/dream/varplot.py                          158    158     0%
bumps/dream/views.py                            171    171     0%
bumps/dream/walk.py                              47     47     0%
bumps/errplot.py                                 36     36     0%
bumps/fitproblem.py                             342    342     0%
bumps/fitservice.py                              65     65     0%
bumps/fitters.py                                641    641     0%
bumps/formatnum.py                              281    281     0%
bumps/gui/__init__.py                             0      0   100%
bumps/gui/about.py                               57     57     0%
bumps/gui/app_frame.py                           94     94     0%
bumps/gui/app_panel.py                          377    377     0%
bumps/gui/console.py                             90     90     0%
bumps/gui/convergence_view.py                    51     51     0%
bumps/gui/data_view.py                          129    129     0%
bumps/gui/fit_dialog.py                         160    160     0%
bumps/gui/fit_thread.py                          92     92     0%
bumps/gui/fit_view.py                            33     33     0%
bumps/gui/gui_app.py                            125    125     0%
bumps/gui/input_list.py                         414    414     0%
bumps/gui/log_view.py                            36     36     0%
bumps/gui/parameter_view.py                     191    191     0%
bumps/gui/plot_view.py                           88     88     0%
bumps/gui/resfiles.py                            47     47     0%
bumps/gui/resources/__init__.py                   3      3     0%
bumps/gui/signal.py                              23     23     0%
bumps/gui/summary_view.py                       118    118     0%
bumps/gui/uncertainty_view.py                    85     85     0%
bumps/gui/util.py                                49     49     0%
bumps/gui/utilities.py                          205    205     0%
bumps/history.py                                 95     95     0%
bumps/initpop.py                                140    140     0%
bumps/lsqerror.py                               193    193     0%
bumps/mapper.py                                 179    179     0%
bumps/monitor.py                                 50     50     0%
bumps/mono.py                                    57     57     0%
bumps/mpfit.py                                  658    658     0%
bumps/mystic/__init__.py                          0      0   100%
bumps/mystic/condition.py                       121    121     0%
bumps/mystic/examples/__init__.py                 0      0   100%
bumps/mystic/examples/circle.py                  46     46     0%
bumps/mystic/examples/corana.py                  30     30     0%
bumps/mystic/examples/decay.py                   15     15     0%
bumps/mystic/examples/model.py                   53     53     0%
bumps/mystic/examples/simple.py                   2      2     0%
bumps/mystic/optimizer/__init__.py                0      0   100%
bumps/mystic/optimizer/de.py                     85     85     0%
bumps/mystic/optimizer/diffev_compat.py          62     62     0%
bumps/mystic/solver.py                           76     76     0%
bumps/mystic/stop.py                            320    320     0%
bumps/mystic/util.py                             32     32     0%
bumps/names.py                                   16     16     0%
bumps/options.py                                154    154     0%
bumps/parameter.py                              892    892     0%
bumps/partemp.py                                156    156     0%
bumps/pdfwrapper.py                             141    141     0%
bumps/plotutil.py                                66     66     0%
bumps/plugin.py                                  16     16     0%
bumps/pmath.py                                    1      1     0%
bumps/pymcfit.py                                 84     84     0%
bumps/pytwalk.py                                397    397     0%
bumps/quasinewton.py                            267    267     0%
bumps/random_lines.py                           141    141     0%
bumps/serialize.py                              204    204     0%
bumps/simplex.py                                178    178     0%
bumps/util.py                                   178    178     0%
bumps/vfs.py                                    299    299     0%
bumps/webview/__init__.py                         0      0   100%
bumps/webview/build_client.py                    30     30     0%
bumps/webview/server/__init__.py                  0      0   100%
bumps/webview/server/__main__.py                  3      3     0%
bumps/webview/server/api.py                     776    776     0%
bumps/webview/server/colors.py                    1      1     0%
bumps/webview/server/corrplot.py                153    153     0%
bumps/webview/server/custom_plot.py              32     32     0%
bumps/webview/server/fit_thread.py              122    122     0%
bumps/webview/server/logger.py                   16     16     0%
bumps/webview/server/names.py                     2      2     0%
bumps/webview/server/persistent_settings.py      21     21     0%
bumps/webview/server/state.py                    59     59     0%
bumps/webview/server/state_hdf5_backed.py       477    477     0%
bumps/webview/server/traceplot.py                15     15     0%
bumps/webview/server/varplot.py                 121    121     0%
bumps/webview/server/webserver.py               277    277     0%
bumps/wsolve.py                                 152     66    57%
-----------------------------------------------------------------
TOTAL                                         15974  15881     1%

============================================================================================= short test summary info =============================================================================================
FAILED bumps/wsolve.py::test - AttributeError: module 'scipy.ndimage._delegators' has no attribute '@py_builtins_signature'. Did you mean: 'grey_dilation_signature'?
===================================================================================== 1 failed, 1 passed, 3 warnings in 5.17s =====================================================================================
```

--- a/scipy/ndimage/_ndimage_api.py
+++ b/scipy/ndimage/_ndimage_api.py
@@ -12,4 +12,5 @@ from ._interpolation import *   # noqa: F403
 from ._measurements import *   # noqa: F403
 from ._morphology import *   # noqa: F403
 
-__all__ = [s for s in dir() if not s.startswith('_')]
+# '@' due to pytest bug, scipy/scipy#22236
+__all__ = [s for s in dir() if not s.startswith(('_', '@'))]

